frb_cors/bin/server.dart

```diff
@@ -0,0 +1,120 @@
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:shelf/shelf.dart';
+import 'package:shelf/shelf_io.dart';
+import 'package:shelf_static/shelf_static.dart';
+import 'package:args/args.dart';
+
+Future<String> system(
+  String command,
+  List<String> arguments, {
+  String? pwd,
+  Map<String, String>? env,
+  bool shell = true,
+}) async {
+  print('> $command ${arguments.join(' ')}');
+  final process = await Process.start(
+    command,
+    arguments,
+    runInShell: shell,
+    workingDirectory: pwd,
+    environment: env,
+  );
+  final ret = <String>[];
+  final err = <String>[];
+  process.stdout.transform(utf8.decoder).listen((line) {
+    stdout.write(line);
+    ret.add(line);
+  });
+  process.stderr.transform(utf8.decoder).listen((line) {
+    stderr.write(line);
+    err.add(line);
+  });
+  final exitCode = await process.exitCode;
+  if (exitCode != 0) {
+    throw ProcessException(command, arguments, err.join(''), exitCode);
+  }
+  return ret.join('');
+}
+
+void bail(bool condition, [String? message]) {
+  if (!condition) {
+    if (message != null) print('Error: $message');
+    exit(1);
+  }
+}
+
+void main(List<String> args) async {
+  final parser = ArgParser()
+    ..addSeparator('frb_cors: A simple static file server.')
+    ..addOption('port',
+        abbr: 'p', help: 'HTTP port to listen to', defaultsTo: '8080')
+    ..addOption('root',
+        abbr: 'r', help: 'Root of the Flutter output', defaultsTo: 'build/web')
+    ..addOption('crate',
+        abbr: 'c', help: 'Directory of the crate', defaultsTo: 'native')
+    ..addFlag('release', help: 'Compile in release mode')
+    ..addFlag('help', abbr: 'h');
+  final config = parser.parse(args);
+  if (config['help']) {
+    print(parser.usage);
+    return;
+  }
+
+  final crateDir = config['crate'];
+  bail(await File('$crateDir/Cargo.toml').exists(),
+      '$crateDir is not a crate dir.');
+  final crateName = await system(
+    'sh',
+    ['-c', "cargo read-manifest | jq -r '.name'"],
+    pwd: crateDir,
+    shell: false,
+  );
+  bail(crateName.trim().isNotEmpty, 'Crate name cannot be empty.');
+  await system(
+    'wasm-pack',
+    [
+      'build', '-t', 'no-modules', '-d', '../web/pkg', '--no-typescript',
+      if (!config['release']) '--dev', '.',
+      '--', // cargo build args
+      '-Z', 'build-std=std,panic_abort'
+    ],
+    pwd: crateDir,
+    env: {
+      'FRB_JS': 'pkg/${crateName.trim()}',
+      'RUSTUP_TOOLCHAIN': 'nightly',
```

Comment:
👀 Is nightly a must? Not sure whether people in production will be afraid of this

Reply:
I'm not too worried about this, as WASM in Rust is still very experimental and needs a lot of polishing. Especially since our use-cases revolve around sharing memory between workers.